src/org/oidc/service/base/ServiceContext.java

```diff
@@ -0,0 +1,223 @@
+     */
+    private RegistrationResponse registrationResponse;
+    /**
+     * A client will need to dynamically construct a number of URLs, which
+     * is the basis for all the URLs.
+     */
+    private String baseUrl;
+    /**
+     * Doing Authorization request parts or the whole request can be
+     * passed by reference using the requestUri. The request itself
+     * must be stored somewhere hence the requestsDirectory.
+     */
+    private String requestsDirectory;
+    /**
+     * Divergence from the standard can be more or less severe.
+     * Less severe cases can be allowed but only if it’s deemed OK.
+     * The only example of this right now is issuer mismatch.
+     * As an example:
+     * According to the OpenID Connect standard the ‘issuer’ returned
+     * in dynamic discovery must be the same as the value of the ‘iss’
+     * parameter in the Id Token and the discovery URL that was used
+     * without the .well-known part.
+     * See for instance http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation
+     * Some OP implementations doesn’t live up to this.
+     * And that by design. So, if you want to talk to those you have to
+     * allow them to diverge from the standard.
+     */
+    private Map<String,Boolean> allow;
+    /**
+     * If manual client registration is done, this is where the results of
+     * that is kept.
+     * If dynamic client registration is done, this is the result of
+     * mapping the registration response against the clientPreferences
+     */
+    private RegistrationResponse behavior;
+    /**
+     * If dynamic client registration is done, this is where it’s specified
+     * what the client would like to use. This is the basis for the
+     * registration request.
+     */
+    private RegistrationRequest clientPreferences;
+    /**
+     * The client identifier, which is always required.
+     */
+    private String clientId;
+    /**
+     * The client secret, which is optional.
+     */
+    private String clientSecret;
+    /**
+     * The Issuer ID. This is the unique identifier of the OP/AS the
+     * client is communicating with.
+     */
+    private String issuer;
+    /**
+     * An unordered list of redirect uris that the client expects to use.
+     */
+    private List<String> redirectUris;
+    /**
+     * redirectUris contains a list of unspecified redirect uris.
+     * In reality there are good reasons for having separated redirect
+     * uris for different expected response modes.
+     * At this time we know of 3 such modes:
+     * fragment, queryString, formPost
+     * In callback, we can keep the redirect uris per response mode
+     * separate.
+     */
+    private Map<DataLocation,String> callBack;
+
+    public ServiceContext(
+            KeyJar keyJar,
+            ServiceContextConfig config) {
+        this.keyJar = keyJar;
+        this.config = config;
+    }
+
+    public ServiceContext() {
```

Comment:
Don't you need to initialize config and keyjar?

Reply:
this is used for testing purposes